helm upgrade --dependency-update panics on OCI chart dependencies (nil RegistryClient)

### What happened?

`helm upgrade --dependency-update` panics with a nil-pointer dereference when the
chart declares an OCI dependency whose version is a range (not an explicit
version).

The upgrade command builds its `downloader.Manager` without a `RegistryClient`
field (`pkg/cmd/upgrade.go`), unlike the identical literals in `install.go`,
`dependency_update.go`, and `dependency_build.go`, which all set it. Resolving an
OCI dependency with a non-explicit version calls
`internal/resolver/resolver.go:151` → `registry.(*Client).Tags`, whose body
dereferences the nil receiver at `pkg/registry/client.go`, crashing the process.

`helm install --dependency-update` works; only `upgrade` regresses.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x78 ...]
helm.sh/helm/v4/pkg/registry.(*Client).Tags(0x0, ...)
	pkg/registry/client.go:792
helm.sh/helm/v4/internal/resolver.(*Resolver).Resolve(...)
	internal/resolver/resolver.go:151
helm.sh/helm/v4/pkg/downloader.(*Manager).resolve(...)
	pkg/downloader/manager.go:240
helm.sh/helm/v4/pkg/downloader.(*Manager).Update(...)
	pkg/downloader/manager.go:199
helm.sh/helm/v4/pkg/cmd.newUpgradeCmd.func2(...)
	pkg/cmd/upgrade.go:230
```

### What did you expect to happen?

`helm upgrade --dependency-update` should resolve the OCI dependency (as
`helm install --dependency-update` does), or fail with a normal error message if
the registry is unreachable. It should never panic.

### How can we reproduce it (as minimally and precisely as possible)?

The crash happens in dependency resolution, before any network call, so no
running registry is required.

1. Create a parent chart `parent/Chart.yaml` with an unresolved OCI dependency
   (a version range, and no matching chart under `parent/charts/`):

   ```yaml
   apiVersion: v2
   name: parent
   version: 0.1.0
   dependencies:
     - name: demo
       repository: oci://localhost:5001/charts
       version: "^1.0.0"
   ```

2. Run:

   ```console
   $ helm upgrade panicdemo ./parent --dependency-update
   ...
   panic: runtime error: invalid memory address or nil pointer dereference
   ```

### Helm version

<details>

```console
$ helm version
version.BuildInfo{Version:"v4.2", GoVersion:"go1.26.5", KubeClientVersion:"v1.36"}
# reproduces on main; the buggy Manager literal is still present on main
```

</details>

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: v1.32.1
Server Version: v1.35.0 (kind)
# not relevant: the panic occurs during dependency resolution, before any
# cluster interaction
```

</details>

--- a/pkg/cmd/upgrade.go
+++ b/pkg/cmd/upgrade.go
@@ -226,6 +226,7 @@ func newUpgradeCmd(cfg *action.Configuration, out io.Writer) *cobra.Command {
 						ContentCache:     settings.ContentCache,
 						Debug:            settings.Debug,
 						SourceDateEpoch:  sourceDateEpoch,
+						RegistryClient:   registryClient,
 					}
 					if err := man.Update(); err != nil {
 						return err

--- a/pkg/cmd/upgrade_test.go
+++ b/pkg/cmd/upgrade_test.go
@@ -18,6 +18,8 @@ package cmd
 
 import (
 	"fmt"
+	"net/http"
+	"net/http/httptest"
 	"os"
 	"path/filepath"
 	"reflect"
@@ -191,6 +193,64 @@ func TestUpgradeCmd(t *testing.T) {
 		},
 	}
 	runTestCmd(t, tests)
+}
+
+// TestUpgradeDependencyUpdateOCINoPanic is a regression test for a nil-pointer
+// panic in `helm upgrade --dependency-update` when a chart declares an OCI
+// dependency. The upgrade command built its downloader.Manager without a
+// RegistryClient (unlike install, dependency update, and dependency build), so
+// resolving an OCI dependency dereferenced a nil *registry.Client. The command
+// must now return a graceful error instead of panicking.
+func TestUpgradeDependencyUpdateOCINoPanic(t *testing.T) {
+	defer resetEnv()()
+
+	// A stub registry that answers the API-version ping but rejects the tag
+	// lookup, so OCI dependency resolution fails fast and hermetically instead
+	// of reaching a real registry.
+	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Docker-Distribution-API-Version", "registry/2.0")
+		if r.URL.Path == "/v2/" {
+			w.WriteHeader(http.StatusOK)
+			return
+		}
+		w.WriteHeader(http.StatusNotFound)
+	}))
+	defer srv.Close()
+
+	// A chart with an unresolved OCI dependency forces --dependency-update into
+	// the tag-lookup path that previously panicked: the version is a range (an
+	// explicit version would skip the lookup) and the dependency is not present
+	// under charts/.
+	tmp := t.TempDir()
+	parent := &chart.Chart{
+		Metadata: &chart.Metadata{
+			APIVersion: chart.APIVersionV2,
+			Name:       "oci-parent",
+			Version:    "0.1.0",
+			Dependencies: []*chart.Dependency{{
+				Name:       "subchart",
+				Repository: fmt.Sprintf("oci://%s/charts", srv.Listener.Addr()),
+				Version:    "^1.0.0",
+			}},
+		},
+	}
+	if err := chartutil.SaveDir(parent, tmp); err != nil {
+		t.Fatalf("Error creating chart: %v", err)
+	}
+	chartPath := filepath.Join(tmp, parent.Metadata.Name)
+	// SaveDir writes only resolved subcharts (Chart.Dependencies()), not the
+	// declared Metadata.Dependencies, so create the empty charts/ directory
+	// explicitly to make the "dependency missing from charts/" state concrete.
+	if err := os.MkdirAll(filepath.Join(chartPath, "charts"), 0o755); err != nil {
+		t.Fatalf("Error creating charts dir: %v", err)
+	}
+
+	// The command must return an error (registry rejects the lookup), not panic.
+	_, _, err := executeActionCommandC(storageFixture(),
+		fmt.Sprintf("upgrade --dependency-update --plain-http oci-parent '%s'", chartPath))
+	if err == nil {
+		t.Fatal("expected an error resolving the OCI dependency, got nil")
+	}
 }
 
 func TestUpgradeWithValue(t *testing.T) {

